sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StateStoreCoordinator.scala

```diff
@@ -164,13 +289,157 @@ private class StateStoreCoordinator(override val rpcEnv: RpcEnv)
+                    baseLogMessage + log", latest snapshot: no upload for query run)"
+                }
+                logWarning(logMessage)
+              }
+          }
+        }
+      }
+      context.reply(true)
+
+    case GetLatestSnapshotVersionForTesting(providerId) =>
+      val version = stateStoreLatestUploadedSnapshot.get(providerId).map(_.version)
+      logDebug(s"Got latest snapshot version of the state store $providerId: $version")
+      context.reply(version)
+
+    case GetLaggingStoresForTesting(queryRunId, latestVersion) =>
+      val currentTimestamp = System.currentTimeMillis()
+      // Only report if the corresponding run has all the necessary information to start reporting
+      if (shouldCoordinatorReportSnapshotLag(queryRunId, latestVersion, currentTimestamp)) {
+        val laggingStores = findLaggingStores(queryRunId, latestVersion, currentTimestamp)
+        logDebug(s"Got lagging state stores: ${laggingStores.mkString(", ")}")
+        context.reply(laggingStores)
+      } else {
+        context.reply(Seq.empty)
+      }
+
     case StopCoordinator =>
       stop() // Stop before replying to ensure that endpoint name has been deregistered
       logInfo("StateStoreCoordinator stopped")
       context.reply(true)
   }
+
+  private def findLaggingStores(
+      queryRunId: UUID,
+      referenceVersion: Long,
+      referenceTimestamp: Long): Seq[StateStoreProviderId] = {
+    // Determine alert thresholds from configurations for both time and version differences.
+    val snapshotVersionDeltaMultiplier =
+      sqlConf.stateStoreCoordinatorMultiplierForMinVersionDiffToLog
+    val maintenanceIntervalMultiplier = sqlConf.stateStoreCoordinatorMultiplierForMinTimeDiffToLog
+    val minDeltasForSnapshot = sqlConf.stateStoreMinDeltasForSnapshot
+    val maintenanceInterval = sqlConf.streamingMaintenanceInterval
+
+    // Use the configured multipliers to determine the proper alert thresholds
+    val minVersionDeltaForLogging = snapshotVersionDeltaMultiplier * minDeltasForSnapshot
+    val minTimeDeltaForLogging = maintenanceIntervalMultiplier * maintenanceInterval
+
+    // Look for active state store providers that are lagging behind in snapshot uploads
+    instances.keys.filter { storeProviderId =>
+      // Only consider providers that are part of this specific query run
+      val latestSnapshot = stateStoreLatestUploadedSnapshot.getOrElse(
+        storeProviderId,
+        defaultSnapshotUploadEvent
+      )
+      storeProviderId.queryRunId == queryRunId && (
+        // Mark a state store as lagging if it's behind in both version and time.
+        // A state store is considered lagging if it's behind in both version and time according
+        // to the configured thresholds.
+        // Stores that didn't upload a snapshot will be treated as a store with a snapshot of
+        // version 0.
+        referenceVersion - Math.max(latestSnapshot.version, 0) > minVersionDeltaForLogging &&
```

Comment:
BTW, you can make the version for the default 0, instead of -1. You won't need this then